### PR TITLE
[canonical_serialzier] Implement a more generic api for serializing maps

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

Currently when serializing a BTreeMap, we are just sorting on the serialized result of the key. This means that we really don't rely on the fact that the key is sorted. As a result, we can relax the api a bit so that it will just take a iterator of key-value tuples and serialize them.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?
Yes

## Test Plan

cargo test

The test suite is going to make sure that we have exactly the same semantics as before.

### CODE REVIEW

File: common/canonical_serialization/src/simple_serializer.rs [hunk 1]
```diff
@@ -58,29 +58,33 @@ where
...
+    {
+        let mut map = BTreeMap::new();
+        for (key, value) in iter {
+            map.insert(
+                SimpleSerializer::<Vec<u8>>::serialize(&key)?,
+                SimpleSerializer::<Vec<u8>>::serialize(&value)?,
+            );
+        }
+
         ensure!(
```
Comment (runtian-zhou) [inline common/canonical_serialization/src/simple_serializer.rs]: Are we changing the semantics here? Should I add the length check at each round of iteration?

File: common/canonical_serialization/src/simple_serializer.rs [hunk 2]
```diff
@@ -58,29 +58,33 @@ where
...
-            v.len(),
+            map.len(),
             ARRAY_MAX_LENGTH,
         );
 
         // add the number of pairs in the map
-        self.encode_u32(v.len() as u32)?;
+        self.encode_u32(map.len() as u32)?;
 
         // Regardless of the order defined for K of the map, write in the order of the lexicographic
```
Comment (davidiw) [inline common/canonical_serialization/src/simple_serializer.rs]: I think these belong before line 68

File: common/canonical_serialization/src/simple_serializer.rs (hunk 1 above)
Comment (davidiw) [reply common/canonical_serialization/src/simple_serializer.rs]: No, this is purely the number of elements and not the total space occupied by the map.

File: common/canonical_serialization/src/canonical_serialize.rs [hunk 3]
```diff
@@ -64,7 +64,16 @@ pub trait CanonicalSerializer {
...
         &mut self,
         v: &BTreeMap<K, V>,
-    ) -> Result<&mut Self>;
+    ) -> Result<&mut Self> {
+        self.encode_tuple_iterator(v.iter())
+    }
+
+    fn encode_tuple_iterator<K: CanonicalSerialize, V: CanonicalSerialize, I>(
+        &mut self,
+        iter: I,
```
Comment (davidiw) [inline common/canonical_serialization/src/canonical_serialize.rs]: Uhh, this stinks, I really miss OO code...

**Comment by davidiw** [comment]
Actually needs a fix before landing...

File: common/canonical_serialization/src/canonical_serialize.rs (hunk 3 above)
Comment (wqfish) [reply common/canonical_serialization/src/canonical_serialize.rs]: If `iter: impl Iterator<Item = (K, V)>` works this might look a bit better.

**Comment by bmwill** [comment]
Is there a reason why we couldn't implement a Serde Serializer/Deserializer for LCS?

File: common/canonical_serialization/src/simple_serializer.rs (hunk 2 above)
Comment (runtian-zhou) [reply common/canonical_serialization/src/simple_serializer.rs]: We can't do this unfortunately. The problem here is that iterator doesn't have a size. Size can only be derived when we finished reconstructing the map.

**Comment by runtian-zhou** [comment]
Sorry what was the fix that I missed?

**Pull request closed without merging**